buildtools/fslex/fslexdriver.fs

```diff
@@ -147,56 +176,82 @@ let writeUnicodeTranslationArray dfaNodes domain (writer: Writer) =
                 let c = char i
                 emit (EncodeChar c parseContext)
             emit Eof
-            writer.writeLine "|];"
+            writer.WriteLine "|];"
         done
 
-    writer.writeLine "    |] "
+    writer.WriteLine "    |] "
 
 let writeUnicodeActionsArray dfaNodes (writer: Writer) =
-    writer.write "let actions : uint16[] = [|"
+    writer.Write "let actions : uint16[] = [|"
     for state in dfaNodes do
         if state.Accepted.Length > 0 then
           writer.WriteUint16 (snd state.Accepted.Head)
         else
           writer.WriteUint16 sentinel
     done
-    writer.writeLine  "|]"
+    writer.WriteLine  "|]"
 
 let writeUnicodeTables lexerLibraryName domain dfaNodes (writer: Writer) =
     writeUnicodeTranslationArray dfaNodes domain writer
     writeUnicodeActionsArray dfaNodes writer
-    writer.writeLine  "let _fslex_tables = %s.%sTables.Create(trans,actions)" lexerLibraryName (match domain with | Unicode -> "Unicode" | ASCII -> "Ascii")
+    writer.WriteLine  "let _fslex_tables = %s.%sTables.Create(trans,actions)" lexerLibraryName (match domain with | Unicode -> "Unicode" | ASCII -> "Ascii")
 
 let writeRules (rules: Rule list) (perRuleData: PerRuleData) outputFileName (writer: Writer) =
-    writer.writeLine  "let rec _fslex_dummy () = _fslex_dummy() "
+    writer.WriteLine  "let rec _fslex_dummy () = _fslex_dummy() "
 
     // These actions push the additional start state and come first, because they are then typically inlined into later
     // rules. This means more tailcalls are taken as direct branches, increasing efficiency and
     // improving stack usage on platforms that do not take tailcalls.
     for (startNode, actions),(ident,args,_) in List.zip perRuleData rules do
-        writer.writeLine "// Rule %s" ident
-        writer.writeLine "and %s %s lexbuf =" ident (String.Join(" ", Array.ofList args))
-        writer.writeLine "  match _fslex_tables.Interpret(%d,lexbuf) with" startNode.Id
+        writer.WriteLine "// Rule %s" ident
+        writer.WriteLineInterface "/// Rule %s" ident
+        let arguments =
+            args
+            |> List.map (function
+                | RuleArgument.Ident ident -> ident
+                | RuleArgument.Typed(ident, typ) -> sprintf "(%s: %s)" ident typ)
+            |> String.concat " "
+
+        writer.WriteLine "and %s %s lexbuf =" ident arguments
+        
+        let signature =
+            if List.isEmpty args then
+                sprintf "val %s: lexbuf: LexBuffer<char> -> token" ident
+            else
+                args
+                |> List.map (function
+                    | RuleArgument.Ident ident ->
+                        // This is not going to lead to a valid signature file, the only workaround is that the caller will specify the type.
```

Comment:
What does it mean in practise that it is not valid?
Do we have any issues with current code?

Reply:
If you didn't add a type annotation in the `*.fsl` file, the generated signature file becomes invalid and you cannot use it in your project. The existing `*.fs` will not be affected by this.